### ISSUE TITLE
Quickly specify parameters under a route

### ISSUE THREAD

**Issue by clawconduce** [go-swagger/go-swagger#782]
## Problem statement

I would like to specify parameters when creating a route, similar to Consumes, Produces, Responses, etc.  These details are the first that come to mind, so it could probably use some iteration.

```
// swagger:route PUT /profile/{id} update_profile
//
// Update a user profile
//
//     Parameters:
//      - id: [in:path required:true type:string] Description goes here
//      - profile: [in:body required:true model:profile_model]
//     Consumes:
//     - application/json
//     Produces:
//     - application/json
//     Responses:
//       200: ok
//       404: NotFound
```

In other places:
```
// swagger:route GET /profile/{id} get_profile
//
// Get a user profile
//
//     Parameters:
//      - id: [in:path required:true] Description goes here
//     Consumes:
//     - application/json
//     Produces:
//     - application/json
//     Responses:
//       200: profile_model
```
```
// swagger:model profile_model
type Profile struct {
    //a bunch of model info here
}
```

And this fits with #781, it would be nice to create a parameter and reuse it, so for example
```
// swagger:parameter profile_id_param
type ProfileIDParam struct {
    /*Profile id
      Required: true
      In: path
    */
    ID string `json:"id"`
}
```

and in the examples above, replace the lines that specify id with something like:
```
// Parameters
//  - predefined:profile_id_param
```

**Comment by Sakibs** [comment]
I like this idea. I also submitted a similar request
https://github.com/go-swagger/go-swagger/issues/779

It would also be great if we could add headers as well

**Comment by casualjim** [comment]
You can do much of what you want to achieve today.
Parameter structs can be associated with more than one operation, and you can define more than one parameter struct.

```markdown
swagger:params [operationid1 operationid2]

Links a struct to one or more operations. The params in the resulting swagger spec 
can be composed of several structs.  There are no guarantees given on how property 
name overlaps are resolved when several structs apply to the same operation.
```

So the idea is that when you want to have a parameter struct that is comprised out of several parameter struct, you don't annotate the one you bind to but rather compose it out of other pieces.

Then the second ask for the responses, there is a syntax for that use case.

```go
// swagger:route PUT /orders/{id} orders updateOrder
//
// Responses:
// default: body:genericError
// 200: body:someResponse
// 422: body:validationError
mountItem("PUT", basePath+"/orders/:id", nil)


// A GenericError is the default error message that is generated.
// For certain status codes there are more appropriate error structures.
//
// swagger:response genericError
type GenericError struct {
	// in: body
	Body struct {
		Code    int32 `json:"code"`
		Message error `json:"message"`
	} `json:"body"`
}

// A ValidationError is an that is generated for validation failures.
// It has the same fields as a generic error but adds a Field property.
//
// swagger:response validationError
type ValidationError struct {
	// in: body
	Body struct {
		Code    int32  `json:"code"`
		Message string `json:"message"`
		Field   string `json:"field"`
	} `json:"body"`
}

// A ModelAndResponse contains a params and model annotation, so it's invalid
//
// swagger:model theModelName
// swagger:response someResponse
type ModelAndResponse struct {
	ID int64 `json:"id,omitempty"`
}

```

**Comment by clawconduce** [comment]
This almost solves my problem - my issue is that I am generating the models in a previous run of swagger and passing that json file as an input, and so I do not have model objects.  This strange setup is a side effect of using protobuf & the probably also because I'm still learning how to use go-swagger.  I'm going to give it a 2nd attempt and see if I can clean it up.

However, I still have the original request.  I would like to create parameters under a route without having to create a struct with a `swagger:parameters` comment.  As mentioned in #781, it would be nice to point to existing parameters here too, but that's a very soft request.

**Comment by GlenDC** [comment]
@casualjim the only thing I don't understand is why everything has to be linked to a struct? Why can't we simply define it as a document? For example I don't want a params struct, yet i want to be able to define the (PATH) params for a certain route? Right now that's only be possible when using a struct, no?

Would it be going against the design of go-swagger if I would work on a PR for this, so that we can define params and responses all from within the route comments? Rather than having to link them to structs?

Also is there a way to define body response structs without the need for that extra Body struct value? It's a bit weird to have it there from a golang user perspective.

**Comment by casualjim** [comment]
What syntax would you propose for this?

**Comment by clawconduce** [comment]
@GlenDC What about `200: body:modelname`?  This is under the `swagger:route` comment though, so it may not solve your problem.

I am adding a little parsing for the responses to support a `description:` tag.  Feel free to use that, or come up with a different way.  In general I like the idea of having `tag1:value1 tag2:value2`, etc, but this doesn't fit with descriptions, which I think is safe to assume have spaces in the value.  My workaround is to make everything after `desription:` part of the description.  See #779 for more detail.

**Comment by casualjim** [comment]
what is the value of adding more comments, because you can deep merge an existing swagger file. Doesn't that solve the problem for you @GlenDC. The benefit is that you can use swagger validation to validate your partial spec

**Comment by GlenDC** [comment]
Maybe what I'm trying to achieve is a bit idiomatic. But the value would be that I wouldn't have a swagger file ever, and always have it generated automatically purely on the server codebase, such that the documentation, system tests and client code is always directly generated from the generate spec, which is generated on itself from the server code.

For the most part you want to have your spec generated from golang related code, such as response structs. But sometimes you're returning a json simply as a `map[string]interface{}`. Right now that doesn't seem to be possible.

@clawconduce the description tag is a beginning. But if you look at how one would normally write a fully manual openapi (swagger 2.0) spec file, there is so much more you can give it. For the cases where you would want that, because you wouldn't have access to a struct, it seems like a useful thing to have.

So to recap. I agree that you want to keep the spec comments as closely linked to actual Golang code as possible. But, because there will always be situations where that is not possible or desired, it should be possible to do anything the specs allows via comments alone. Shouldn't be too hard to achieve, should it? And I don't want to have some bits in a spec file, as the entire goal is to keep it all directly linked from within the code files, just as you also want to generate normal documentation directly from in-line comments. (think Godocs, Rust Docs)

**Comment by GlenDC** [comment]
And I guess the syntax would be as close as possible to the default YAML syntax, I don't see why you would go away from that, seems already simple enough. And makes it easier to learn to use it, as people who want to use go-swagger, probably already know swagger/openapi and its spec.

**Comment by clawconduce** [comment]
I agree with most of what @GlenDC says - in my situation, I would like the comments as close as possible to the code so it's right in your face when you're updating code.  I do like the yaml format, but I shied away from that because it is a lot of parsing work.

I do have a valid use case for the input swagger, but it's a little strange.  We migrated most of a python codebase to golang, but there is still some python code left behind it is complex & not easy to port.  Thanks to the magic of nginx, we can hide this fact from anyone using the API.  So I am using a swagger file for the Python endpoints as an input.

I also use the input swagger a few places, I could probably move to the code.  Our models are defined in protobuf, and the code for structs is built from that.  I put example values in the swaggerfile.  I also have some responses that are used in many places, but I didn't want to add a struct for - these are all 4xx responses.  I could use the #779 solution, but I wanted to point to a response (`#/responses/`) so the descriptions all came from the same place.  I think both could be solved by using structs, but I'd prefer just to write yaml/json in comments (I say json because for big json model examples, I wouldn't want to translate the example to yaml).

**Comment by casualjim** [comment]
so you want to provide a swagger spec like the input spec but in the code?

**Comment by GlenDC** [comment]
The parts where it's needed yes, the part where it isn't needed I more than happily link directly to Golang code. But having this alternative for when I need it would definitely be great. I don't mind putting some time in to make this happen with a PR. Would just need some pointers to get me started and know of any gotchas, as I would be new to this project and codebase.

Some responses could be linked to a struct, as it exists. But because currently you seem to need to put your struct value members within a body struct within the actual struct, I can't use that method anyway.

Another obvious thing that seems to be missing is adding other information to routes such as examples. As that's also kind of information that would be needed in order to automatically run system (end-to-end) tests based on a spec file.

**Comment by casualjim** [comment]
So you want to provide a value you parse out of the comments

it would probably be injected here:
https://github.com/go-swagger/go-swagger/blob/master/scan/scanner.go#L192

Now as to implementation of it, it would be something like a `swagger:input` annotation (which is new) and then the rest of that comment block can be considered a swagger specification
If the first char is { parse json otherwise try yaml

**Comment by GlenDC** [comment]
Cool, sounds like a reasonable approach. I must say I'm a bit surprised of having a new `swagger:input` annotation proposal, as I'm not planning to write an entire specification, as one comment, as that gets a bit crazy, perhaps?

Was more thinking to be able to provide a complete openapi router specification, under the `swagger:router` for a router, so being able to provide parameters, responses, examples and whatnot under that annotation. Or are you saying that the `swagger:input` should also be able to handle such incomplete specifications (e.g. just 1 router definition? As I do still want the documentation of each router right above the actual router creation. Does that make sense?

**Comment by casualjim** [comment]
oic I was coming from the POV of having a replacement for the input swagger specification instead of what you describe.
In case of the swagger:router I think what you want is to just a few more keys that are recognized, in that case you can extend what exists today for the swagger:router.
A good starting point to read up on that is: https://github.com/go-swagger/go-swagger/blob/master/scan/routes.go#L194-L199

**Comment by GlenDC** [comment]
After some thought, this issue might be solved with the solution potentially provided in enhancement issue #793

**Comment by clawconduce** [comment]
Yes, it would!  I was doing searches for keywords and never came across it.  @casualjim @GlenDC Should this be closed in favor of #793?

**Comment by casualjim** [comment]
superseded by: #793
